File: server/application.py
"""Web server to manage commits and pushes"""
import json
import logging
import shutil
import subprocess
import time
from os.path import basename
from pathlib import Path
from zipfile import ZipFile

from flask import Flask, abort, request
from flask_cors import CORS

from server.diff import DiffGen

app = Flask(__name__)
CORS(app)

log = logging.getLogger("werkzeug")
log.disabled = True


class ProjectConfig:
    def __init__(self, file_name: str) -> None:
        self.file_name = file_name
        if Path(file_name).exists():
            with open(Path(file_name), encoding="utf-8") as fh:
                self.map = json.load(fh)
        else:
            Path(file_name).write_text("{}", encoding="utf-8")
            self.map = {}

    def save(self) -> None:
        Path(self.file_name).write_text(json.dumps(self.map), encoding="utf-8")


Path("projects").mkdir(exist_ok=True)
config = ProjectConfig("projects/project_config.json")


def extract_project(project_name: str) -> None:
    project = config.map[project_name]

    with ZipFile(project["project_file"], "r") as fh:
        fh.extractall(project["base"])


@app.get("/create_project")
def create_project():  # type: ignore
    file_name = request.args.get("file_name")
    if file_name is None:
        abort(500)
    name = basename(Path(file_name)).split(".")[0]

    project_path = ("projects" / Path(name)).absolute()
    file_path = Path(file_name).absolute()

    if not config.map.get(name):
        config.map[name] = {
            "base": str(project_path),
            "project_file": str(file_path),
        }
    else:
        i = 0
        if Path(f"{name}~0").exists():
            while Path(f"{name}~{i}").exists():
                i += 1
        name = f"{name}~{i}"
        config.map[name] = {
            "base": str(Path(name).absolute()),
            "project_file": str(file_path),
        }

    config.save()
    if not project_path.exists():
        project_path.mkdir(parents=True)
    extract_project(basename(project_path))

    init_repo = subprocess.check_output(["git", "init"], cwd=project_path).decode()
    if not "Git repository" in init_repo:
        abort(500)
    if subprocess.check_call(["git", "add", "."], cwd=project_path) != 0:
        abort(500)
    try:
        subprocess.check_call(
            ["git", "commit", "-m", "Initial commit"], cwd=project_path
        )
    except subprocess.CalledProcessError:
        abort(500)

    return {"project_name": name}


@app.post("/find_project")
def find_project():  # type: ignore
    project_json = request.get_json().get("project_json")

    if project_json is None:
        abort(500)

    for f in Path("projects").glob("**/project.json"):
        if project_json == f.read_text():
            return {"project_name": basename(str(f.parent))}

    return {"project_name": None}


@app.get("/<project_name>/unzip")
def unzip(project_name):  # type: ignore
    project = config.map[project_name]
    project_dir = Path(project["base"])

    try:
        shutil.copyfile(project_dir / "project.json", project_dir / "project.old.json")
    except FileNotFoundError:
        pass

    time.sleep(1)
    with ZipFile(project["project_file"], "r") as fh:
        fh.extractall(project_dir)

    return {}


@app.get("/<project_name>/commit")
def commit(project_name):  # type: ignore
    """Commits the state of a Scratch project and returns a message"""
    project_dir = Path(config.map[project_name]["base"])

    with open(project_dir / "project.old.json", encoding="utf-8") as fh:
        current_project = DiffGen(json.load(fh))

    with open(project_dir / "project.json", encoding="utf-8") as fh:
        new_project = DiffGen(json.load(fh))

    costume_removals = new_project.costumes(current_project)
    commit_message = ", ".join(current_project.commits(new_project))

    for _, (path, _) in costume_removals:
        Path(project_dir / path).unlink(missing_ok=True)

    if subprocess.call(["git", "add", "."], cwd=project_dir) != 0:
        abort(500)

    with subprocess.Popen(
        ["git", "commit", "-m", commit_message],
        cwd=project_dir,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    ) as commit_cmd:
        if commit_cmd.returncode != 0:
            _, stderr = commit_cmd.communicate()
            # TODO: modify below code, why do I depend on a commit hook?
            if ".git/hooks/commit-msg" not in stderr.decode("utf-8"):
                abort(500)

    return commit_message


@app.get("/<project_name>/push")
def push(project_name):  # type: ignore
    """Pushes to a remote Git repository"""
    project_dir = Path(config.map[project_name]["base"])

    if subprocess.call(["git", "push"], cwd=project_dir) != 0:
        abort(500)

    return {}


@app.get("/<project_name>/project.old.json")
def old_project(project_name):  # type: ignore
    """Retreive the project.old.json"""
    project_dir = Path(config.map[project_name]["base"])

    with open(project_dir / "project.old.json", encoding="utf-8") as fh:
        project = json.load(fh)
        return [
            e["blocks"]
            for e in project["targets"]
            if e["name"] == request.args.get("name")
        ][0]


@app.get("/<project_name>/project.json")
def current_project_(project_name):  # type: ignore
    """Retreive the current project.json"""
    project_dir = Path(config.map[project_name]["base"])

    with open(project_dir / "project.json", encoding="utf-8") as fh:
        project = json.load(fh)
        return [
            e["blocks"]
            for e in project["targets"]
            if e["name"] == request.args.get("name")
        ][0]


@app.get("/<project_name>/commits")
def commits(project_name):  # type: ignore
    """Retrieve commits"""
    project_dir = Path(config.map[project_name]["base"])

    with subprocess.Popen(
        [
            "git",
            "log",
            "--pretty=format:"
            + '{%n  "commit": "%H",%n  "subject": "%s",%n  "body": "%b",%n  "author": {%n    "name": "%aN",%n    "email": "%aE",%n    "date": "%aD"%n  }%n}',
        ],
        cwd=project_dir,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    ) as commit_cmd:
        output, _ = commit_cmd.communicate()
        output = json.loads(
            "[" + output.decode().replace("  }\n}", "  }\n},")[:-1] + "]", strict=False
        )

    return output


@app.get("/<project_name>/sprites")
def sprites(project_name):  # type: ignore
    """Retreive sprites that have been changed since project changes"""
    project_dir = Path(config.map[project_name]["base"])

    with open(project_dir / "project.old.json", encoding="utf-8") as fh:
        current_project = DiffGen(json.load(fh))

    with open(project_dir / "project.json", encoding="utf-8") as fh:
        new_project = DiffGen(json.load(fh))

    return {
        "sprites": [
            commit.split(":")[0] for commit in current_project.commits(new_project)
        ]
    }

File: src/server/application.rs
use crate::diff::Diff;
use crate::utils::projects::ProjectConfig;
use crate::utils::extract::extract;

use rocket::http::{ContentType, Status};
use rocket::serde::json::Json;
use rocket::serde::Deserialize;
use rocket::{Build, Rocket};

use serde_json::{json, to_string, Value};

use std::fs;
use std::{
    path::{Path, PathBuf},
    process::{Command, Stdio},
    sync::{Mutex, OnceLock},
    thread::sleep,
    time::Duration,
};

#[derive(Deserialize)]
struct DiffParams {
    old_content: String,
    new_content: String,
}

type AppResponse = (Status, (ContentType, String));

fn response(status: Status, content: Value) -> AppResponse {
    (status, (ContentType::JSON, to_string(&content).unwrap()))
}

fn get_project_path(projects: &Value, project_name: &String) -> PathBuf {
    let base_loc = &projects[project_name]["base"].as_str().unwrap().to_string();
    Path::new(&base_loc).to_path_buf()
}

fn project_config() -> &'static Mutex<ProjectConfig> {
    static CONFIG: OnceLock<Mutex<ProjectConfig>> = OnceLock::new();
    CONFIG.get_or_init(|| Mutex::new(ProjectConfig::new("projects/config.json".to_string())))
}

#[post("/diff", format = "json", data = "<diff_params>")]
fn diff(diff_params: Json<DiffParams>) -> AppResponse {
    response(
        Status::Ok,
        json!(crate::utils::git_piping::git_diff(
            diff_params.old_content.clone(),
            diff_params.new_content.clone()
        )),
    )
}

#[get("/create-project?<file_name>")]
fn create_project(file_name: &str) -> AppResponse {
    let binding = Path::new(file_name)
        .file_name()
        .unwrap()
        .to_os_string()
        .to_str()
        .unwrap()
        .to_string();
    let name = binding.split(".").collect::<Vec<_>>()[0].to_string();
    let mut config = project_config().lock().unwrap();

    let project_path_result = fs::canonicalize(Path::new("projects").join(&name));
    let project_path = match project_path_result {
        Ok(file) => file,
        Err(_) => {
            let _ = fs::create_dir(Path::new("projects").join(&name));
            fs::canonicalize(Path::new("projects").join(&name)).unwrap()
        }
    };

    let Ok(file_path) = fs::canonicalize(&file_name) else {
        return response(
            Status::InternalServerError,
            json!({ "project_name": "fail" }),
        );
    };

    match config.projects[&name] {
        Value::Null => {
            config.projects[name.clone()] = json!({
                "base": project_path.to_str().unwrap().to_string(),
                "project_file": file_path.to_str().unwrap().to_string()
            });
        }
        Value::Object(_) => {
            return response(
                Status::InternalServerError,
                json!({ "project_name": "exists" }),
            );
        }
        _ => todo!("idk"),
    };

    if !project_path.exists() {
        let _ = fs::create_dir(&project_path);
    }

    config.save();

    let project_to_extract = &config.projects[project_path
        .file_name()
        .unwrap()
        .to_str()
        .unwrap()
        .to_string()];
    let target_dir = PathBuf::from(Path::new(
        &project_to_extract["base"].as_str().unwrap().to_string(),
    ));
    // let archive = fs::read()
    // .unwrap();
    extract(fs::File::open(Path::new(
        &project_to_extract["project_file"]
            .as_str()
            .unwrap()
            .to_string(),
    )).expect("failed to open file"), target_dir).unwrap();

    let init_repo = Command::new("git")
        .args(["init"])
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .current_dir(&project_path)
        .output()
        .expect("failed to intialize git repo");
    if !String::from_utf8(init_repo.stdout)
        .unwrap()
        .contains("Git repository")
    {
        return response(
            Status::InternalServerError,
            json!({ "project_name": "fail" }),
        );
    }
    if !Command::new("git")
        .args(["add", "."])
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .current_dir(&project_path)
        .status()
        .unwrap()
        .success()
    {
        return response(
            Status::InternalServerError,
            json!({ "project_name": "fail" }),
        );
    }
    if !String::from_utf8(
        Command::new("git")
            .args(["commit", "-m", "Initial commit"])
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .current_dir(project_path)
            .output()
            .unwrap()
            .stdout,
    )
    .unwrap()
    .contains("Initial commit")
    {
        return response(
            Status::InternalServerError,
            json!({ "project_name": "fail" }),
        );
    }

    response(
        Status::Ok,
        json!({ "project_name": name.replace("projects/", "") }),
    )
}

#[get("/<project_name>/unzip")]
fn unzip(project_name: String) -> AppResponse {
    let projects = &project_config().lock().unwrap().projects;
    let pth = get_project_path(projects, &project_name);

    fs::copy(pth.join("project.json"), pth.join("project.old.json"))
        .expect("failed to move project.json");

    sleep(Duration::from_millis(1000));
    let target_dir = PathBuf::from(Path::new(&pth));
    extract(fs::File::open(Path::new(
        &projects[project_name]["project_file"]
            .as_str()
            .unwrap()
            .to_string(),
    )).expect("failed to open file"), target_dir).unwrap();

    response(Status::Ok, json!({ "status": "success" }))
}

#[get("/<project_name>/project.old.json?<name>")]
fn old_project(project_name: String, name: String) -> AppResponse {
    let pth = get_project_path(&project_config().lock().unwrap().projects, &project_name);

    let old_project = serde_json::from_str::<serde_json::Value>(
        &fs::read_to_string(pth.join("project.old.json"))
            .unwrap()
            .as_str(),
    )
    .expect("failed to parse project.old.json");

    let blocks = old_project["targets"]
        .as_array()
        .unwrap()
        .iter()
        .filter(|t| t["name"].as_str().unwrap().to_string() == name)
        .map(|t| t["blocks"].as_object().unwrap())
        .next()
        .unwrap()
        .to_owned();

    response(Status::Ok, serde_json::Value::Object(blocks))
}

#[get("/<project_name>/project.json?<name>")]
fn current_project(project_name: String, name: String) -> AppResponse {
    let pth = get_project_path(&project_config().lock().unwrap().projects, &project_name);

    let old_project = serde_json::from_str::<serde_json::Value>(
        &fs::read_to_string(pth.join("project.json"))
            .unwrap()
            .as_str(),
    )
    .expect("failed to parse project.json");

    let blocks = old_project["targets"]
        .as_array()
        .unwrap()
        .iter()
        .filter(|t| t["name"].as_str().unwrap().to_string() == name)
        .map(|t| t["blocks"].as_object().unwrap())
        .next()
        .unwrap()
        .to_owned();

    response(Status::Ok, serde_json::Value::Object(blocks))
}

#[get("/<project_name>/sprites")]
fn sprites(project_name: String) -> AppResponse {
    let pth = get_project_path(&project_config().lock().unwrap().projects, &project_name);

    let binding = &fs::read_to_string(pth.join("project.old.json"));
    let project_old_json = match binding {
        Ok(fh) => fh.as_str(),
        Err(_) => {
            return response(
                Status::InternalServerError,
                json!({ "status": "you forgot to unzip the project dingus" }),
            )
        }
    };
    let _current_project = serde_json::from_str::<serde_json::Value>(project_old_json).unwrap();

    let current_diff = Diff::new(_current_project);
    let _new_project = serde_json::from_str::<serde_json::Value>(
        &fs::read_to_string(pth.join("project.json"))
            .unwrap()
            .as_str(),
    )
    .unwrap();

    let new_diff = Diff::new(_new_project);

    response(
        Status::Ok,
        json!({
            "sprites": current_diff.commits(&new_diff)
                .iter()
                .map(|commit| commit.split(":").collect::<Vec<_>>()[0])
                .collect::<Vec<_>>()
        }),
    )
}

#[get("/<project_name>/push")]
fn push(project_name: String) -> AppResponse {
    let pth = get_project_path(&project_config().lock().unwrap().projects, &project_name);

    if !Command::new("git")
        .arg("push")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .current_dir(&pth)
        .status()
        .unwrap()
        .success()
    {
        return response(Status::InternalServerError, json!({ "status": "fail" }));
    }

    response(Status::Ok, json!({ "status": "success" }))
}

#[get("/<project_name>/commit")]
fn commit(project_name: String) -> AppResponse {
    let pth = get_project_path(&project_config().lock().unwrap().projects, &project_name);

    let _current_project = serde_json::from_str::<serde_json::Value>(
        &fs::read_to_string(pth.join("project.old.json"))
            .unwrap()
            .as_str(),
    )
    .unwrap();
    let current_diff = Diff::new(_current_project);

    let _new_project = serde_json::from_str::<serde_json::Value>(
        &fs::read_to_string(pth.join("project.json"))
            .unwrap()
            .as_str(),
    )
    .unwrap();
    let new_diff = Diff::new(_new_project);

    let costume_removals = new_diff.costumes(&current_diff);
    let commit_message = current_diff.commits(&new_diff).join(", ");

    for change in costume_removals {
        fs::remove_file(pth.join(change.costume_path)).expect("failed to remove asset");
    }

    if !Command::new("git")
        .args(["add", "."])
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .current_dir(&pth)
        .status()
        .unwrap()
        .success()
    {
        return response(
            Status::InternalServerError,
            json!({ "project_name": "fail" }),
        );
    }

    let mut binding = Command::new("git");
    let git_commit = binding
        .args(["commit", "-m", commit_message.as_str()])
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .current_dir(pth);
    if !git_commit.status().unwrap().success() {
        return response(
            Status::InternalServerError,
            json!({ "project_name": "Nothing to add" }),
        );
    }

    response(Status::Ok, json!({ "message": commit_message }))
}

#[get("/<project_name>/commits")]
fn commits(project_name: String) -> AppResponse {
    let pth = get_project_path(&project_config().lock().unwrap().projects, &project_name);

    let git_log = Command::new("git")
        .args([
            "log",
            &("--pretty=format:".to_owned()+
            "{\"commit\": \"%H\", \"subject\": \"%s\", \"body\": \"%b\", \"author\": {\"name\": \"%aN\", \"email\": \"%aE\", \"date\": \"%aD\"}},")
        ]).current_dir(pth).output().unwrap().stdout;
    let binding = String::from_utf8(git_log).unwrap();
    let binding = if binding.as_str().matches("\"commit\"").count() > 1 {
        format!(
            "[{}]",
            binding.replace(" }\n}", " }\n},").as_str()[..binding.len() - 1].to_string()
        )
    } else {
        format!("[{binding}]")
    };
    let log_output = binding.as_str();

    response(
        Status::Ok,
        serde_json::from_str(("[".to_owned() + &log_output + "]").as_str())
            .expect("failed to parse log output"),
    )
}

pub fn create_app() -> Rocket<Build> {
    let cors = rocket_cors::CorsOptions::default().to_cors();

    rocket::build()
        .mount(
            "/",
            routes![
                diff,
                create_project,
                unzip,
                commit,
                push,
                old_project,
                current_project,
                sprites,
                commits
            ],
        )
        .attach(cors.unwrap())
}
